pennylane/workflow/get_best_diff_method.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains a function for getting the best differentiation method for a given QNode.
+
+"""
+
+from functools import wraps
+
+import pennylane as qml
+from pennylane.workflow.qnode import _make_execution_config
+
+
+# pylint: disable=unused-argument
+def get_best_diff_method(qnode):
+    """Returns a function that computes the 'best' differentiation method
+    for a particular QNode.
+
+    This method attempts to determine support for differentiation
+    methods using the following order:
+
+    * ``"device"``
+    * ``"backprop"``
+    * ``"parameter-shift"``
+    * ``"finite-diff"``
+
+    The first differentiation method that is supported (going from
+    top to bottom) will be returned. Note that the SPSA-based and Hadamard-based gradients
+    are not included here.
+
+    Args:
+        qnode (.QNode): the qnode to get the 'best' differentiation method for.
+        return_as_str (bool): return the 'best' differentiation method in human-readable format.
+
+    Returns:
+        tuple[str or .TransformDispatcher, dict, .device.Device: Tuple containing the ``gradient_fn``,
+        ``gradient_kwargs``, and the device to use when calling the execute function.
```

Comment:
We no longer need to return keyword arguments or the device anymore. Those are historical remnants that are still there because it's hard to change the return value of a public method.  

For some context, when using backpropagation with `default.qubit.legacy` (now relegated to the testing folder), we used to switch between different devices for different interfaces. So we would silently execute on a different device than the one provided.  It led to no end of misery and confusion. Gladly, that is now gone.  We no longer need to switch the device internally.

As for gradient kwargs, I have now added a patch adding the device to the gradient kwargs here:

https://github.com/PennyLaneAI/pennylane/blob/7efa9c4f9f263b49f4bb01195bc9e57f9d68ed00/pennylane/workflow/qnode.py#L890

So we no longer need to add the gradient keyword args in this method.